File: ui-automation-framework/pages/inventory.page.ts
import type { Locator, Page } from '@playwright/test';
import { HeaderComponent } from '../components/header.component';
import { InventoryItemComponent } from '../components/inventory-item.component';

export type SortOption = 'az' | 'za' | 'lohi' | 'hilo';

export class InventoryPage {
  readonly header: HeaderComponent;
  readonly inventoryList: Locator;
  readonly items: Locator;
  readonly itemNames: Locator;
  readonly itemPrices: Locator;
  readonly sortSelect: Locator;

  constructor(readonly page: Page) {
    this.header = new HeaderComponent(page);
    this.inventoryList = page.getByTestId('inventory-list');
    this.items = page.getByTestId('inventory-item');
    this.itemNames = page.getByTestId('inventory-item-name');
    this.itemPrices = page.getByTestId('inventory-item-price');
    this.sortSelect = page.getByTestId('product-sort-container');
  }

  async goto(): Promise<void> {
    await this.page.goto('/inventory.html');
  }

  item(productName: string): InventoryItemComponent {
    const root = this.items.filter({
      has: this.page.getByTestId('inventory-item-name').getByText(productName, { exact: true }),
    });
    return new InventoryItemComponent(root);
  }

  async addToCart(productName: string): Promise<void> {
    await this.item(productName).addToCartButton.click();
  }

  async removeFromCart(productName: string): Promise<void> {
    await this.item(productName).removeButton.click();
  }

  async openProductDetails(productName: string): Promise<void> {
    await this.item(productName).name.click();
  }

  async sortBy(option: SortOption): Promise<void> {
    await this.sortSelect.selectOption(option);
  }

  async visibleNames(): Promise<string[]> {
    return this.itemNames.allInnerTexts();
  }

  async visiblePrices(): Promise<string[]> {
    return this.itemPrices.allInnerTexts();
  }
}

File: ui-tests/src/test/java/saucedemo/pages/InventoryPage.java
package saucedemo.pages;

import com.microsoft.playwright.Locator;
import com.microsoft.playwright.Page;
import saucedemo.components.Header;
import saucedemo.components.ProductItem;

import java.util.ArrayList;
import java.util.List;

public class InventoryPage {
    private final Page page;
    public final Header header;

    public InventoryPage(Page page) {
        this.page = page;
        this.header = new Header(page);
    }

    public void open() {
        page.navigate("/inventory.html");
    }

    public Locator list() {
        return page.locator("[data-test='inventory-list']");
    }

    public Locator items() {
        return page.locator("[data-test='inventory-item']");
    }

    public ProductItem item(String name) {
        for (int i = 0; i < items().count(); i++) {
            Locator one = items().nth(i);
            if (one.locator("[data-test='inventory-item-name']").innerText().equals(name)) {
                return new ProductItem(one);
            }
        }
        throw new RuntimeException("product not found: " + name);
    }

    public void addToCart(String name) {
        item(name).addToCart();
    }

    public void removeFromCart(String name) {
        item(name).removeFromCart();
    }

    public void openProduct(String name) {
        item(name).name().click();
    }

    public void sortBy(String value) {
        // values: az, za, lohi, hilo
        page.locator("[data-test='product-sort-container']").selectOption(value);
    }

    public List<String> visibleNames() {
        List<String> names = new ArrayList<>();
        Locator nameLoc = page.locator("[data-test='inventory-item-name']");
        for (int i = 0; i < nameLoc.count(); i++) {
            names.add(nameLoc.nth(i).innerText());
        }
        return names;
    }

    public List<Double> visiblePrices() {
        List<Double> prices = new ArrayList<>();
        Locator priceLoc = page.locator("[data-test='inventory-item-price']");
        for (int i = 0; i < priceLoc.count(); i++) {
            prices.add(saucedemo.utils.PriceHelper.parse(priceLoc.nth(i).innerText()));
        }
        return prices;
    }
}
